### PR TITLE
Support for multiple subnets in a overlay network

### PR DESCRIPTION
**Pull request by sanimej**
Added the subnet construct in the overlay driver. Bridge and the vxlan dev will be created per subnet.

### CODE REVIEW

File: drivers/overlay/joinleave.go [hunk 1]
```diff
@@ -24,10 +24,25 @@ func (d *driver) Join(nid, eid string, sboxKey string, jinfo driverapi.JoinInfo,
 		return fmt.Errorf("could not find endpoint with id %s", eid)
 	}
 
+	s := n.getSubnetforIP(ep.addr)
+	if s == nil {
+		return fmt.Errorf("could not find subnet for endpoint %s", eid)
+	}
+	// If the subnet doesn't have a vni assigned yet query the datastore first
+	if s.vni == 0 {
```
Comment (mrjana) [inline drivers/overlay/joinleave.go]: Can we just do this check inside `obtainVxlanID` and return success if vni is non-zero rather than checking this in all places that we call `obtainVxlanID`?

File: drivers/overlay/joinleave.go [hunk 2]
```diff
@@ -72,7 +87,7 @@ func (d *driver) Join(nid, eid string, sboxKey string, jinfo driverapi.JoinInfo,
 		}
 	}
 
-	d.peerDbAdd(nid, eid, ep.addr.IP, ep.mac,
+	d.peerDbAdd(nid, eid, ep.addr.IP, ep.addr.Mask, ep.mac,
```
Comment (mrjana) [inline drivers/overlay/joinleave.go]: We can change the peer entry to carry `IPNet` instead of `IP` and `Mask` seperately so that we can always carry them together while sending through serf and while holding it in peerDb

File: drivers/overlay/ov_network.go [hunk 3]
```diff
@@ -10,62 +10,103 @@ import (
...
+			gwIP:     ipd.Gateway,
+			once:     &sync.Once{},
+		}
+		n.subnets = append(n.subnets, s)
+	}
 
-	for i, ipd := range ipV4Data {
-		n.subnets[i] = ipd.Pool
-		n.gateways[i] = ipd.Gateway
+	for i := 0; i < 5; i++ {
```
Comment (mrjana) [inline drivers/overlay/ov_network.go]: Why retry only upto some magic value? I think we retry indefinitely when we get `KeyModified` error but if we do it properly then we should not hit this loop no more that two or three iterations.

File: drivers/overlay/ov_network.go [hunk 4]
```diff
@@ -10,62 +10,103 @@ import (
...
+		if err = n.writeToStore(); err != nil {
+			// On ErrKeyModified retry few times before giving up
+			if err == datastore.ErrKeyModified {
```
Comment (mrjana) [inline drivers/overlay/ov_network.go]: If you don't read the object here you will loop indefinitely(or however many count that you limit) unless you are watching for changes

File: drivers/overlay/ov_network.go [hunk 5]
```diff
@@ -136,14 +188,50 @@ func (n *network) destroySandbox() {
 			iface.Remove()
 		}
 
-		if err := deleteVxlan(n.vxlanName); err != nil {
-			logrus.Warnf("could not cleanup sandbox properly: %v", err)
+		for _, s := range n.subnets {
+			if len(s.vxlanName) != 0 {
```
Comment (mrjana) [inline drivers/overlay/ov_network.go]: You can just say `if s.vxlanName != ""`

File: drivers/overlay/ov_network.go (hunk 3 above)
Comment (sanimej) [inline drivers/overlay/ov_network.go]: yeah, in other places we do retry indefinitely. Agree that normally we shouldn't get the KeyModified error continuously. But as a fail safe mechanism is it good to limit the number of retries ?

File: drivers/overlay/ov_network.go (hunk 4 above)
Comment (sanimej) [inline drivers/overlay/ov_network.go]: good catch..thanks.

File: drivers/overlay/ov_network.go [hunk 6]
```diff
@@ -173,9 +256,11 @@ func (n *network) initSandbox() error {
 	})
 
 	go n.watchMiss(nlSock)
-	return n.initVxlan()
+	return nil
+	//return n.initVxlan()
```
Comment (mrjana) [inline drivers/overlay/ov_network.go]: Can you please remove commented lines?

File: drivers/overlay/ov_network.go [hunk 7]
```diff
@@ -173,9 +256,11 @@ func (n *network) initSandbox() error {
...
 }
 
+/*
```
Comment (mrjana) [inline drivers/overlay/ov_network.go]: Same here. Remove commented lines

File: drivers/overlay/ov_network.go (hunk 6 above)
Comment (sanimej) [inline drivers/overlay/ov_network.go]: @mrjana I left this one intentionally because it wasn't clear to me why this function initVxlan was added. There was only one caller. I have now moved what initVxlan was doing into joinSubnetSandbox(). So if these changes look ok I will remove the commented lines.

File: drivers/overlay/ov_network.go [hunk 8]
```diff
@@ -291,7 +377,19 @@ func (n *network) KeyPrefix() []string {
 }
 
 func (n *network) Value() []byte {
-	b, err := json.Marshal(n.vxlanID())
+	overlayNetmap := make(map[string]interface{})
+
+	s := n.subnets[0]
```
Comment (mrjana) [inline drivers/overlay/ov_network.go]: Are we only taking care of single subnet case for now? That is fine. Just wanted to confirm

File: drivers/overlay/ov_network.go [hunk 9]
```diff
@@ -291,7 +377,19 @@ func (n *network) KeyPrefix() []string {
...
+	if s == nil {
+		//assert here ?
```
Comment (mrjana) [inline drivers/overlay/ov_network.go]: Log an error message here

File: drivers/overlay/ov_network.go (hunk 8 above)
Comment (sanimej) [inline drivers/overlay/ov_network.go]: Yes, I will make the change to handle multiple subnets in a separate PR. Only the marshal/unmarshal needs some changes.

File: drivers/overlay/ov_serf.go [hunk 10]
```diff
@@ -83,7 +83,8 @@ func (d *driver) notifyEvent(event ovNotify) {
 	n := d.network(event.nid)
 	ep := n.endpoint(event.eid)
 
-	ePayload := fmt.Sprintf("%s %s %s", event.action, ep.addr.IP.String(), ep.mac.String())
+	ePayload := fmt.Sprintf("%s %s %s %s", event.action, ep.addr.IP.String(),
```
Comment (mrjana) [inline drivers/overlay/ov_serf.go]: Again I think we probably can send `ep.addr.String()` which will also send the mask and then parse it back to an `IPNet` value in the receiving side rather than sending the `IP` and `Mask` separately.

File: drivers/overlay/ov_network.go (hunk 3 above)
Comment (mrjana) [inline drivers/overlay/ov_network.go]: I don't think we should have an arbitrary number of retries. It's probably fine to retry indefinitely. If at all needed that we retry only a limited number of times we can make a wrapper function to do this job in the future.

File: drivers/overlay/ov_network.go (hunk 6 above)
Comment (mrjana) [inline drivers/overlay/ov_network.go]: Even if there is only one caller sometimes we have a separate function so that the caller function doesn't get too long or get into too many indentation levels. So it is perfectly fine to remove these commented lines

File: drivers/overlay/ov_network.go (hunk 8 above)
Comment (mrjana) [inline drivers/overlay/ov_network.go]: Sure we can do multiple subnets in a different PR.

File: drivers/overlay/ov_network.go (hunk 4 above)
Comment (sanimej) [inline drivers/overlay/ov_network.go]: Thinking about the use case here, we really don't need the AtomicPut because on KeyModified error we are going to read to get the latest index. We can use the PutObject instead. But I think it should be a separate PR.

File: drivers/overlay/ov_network.go (hunk 8 above)
Comment (mavenugo) [inline drivers/overlay/ov_network.go]: why is that ? I thought the primary purpose of this PR is to support multiple subnets. what am i missing ?

Comment (sanimej) [inline drivers/overlay/ov_network.go]: This PR has the required infra to support multiple subnets. I had some issues in marshal/unmarshal of multiple subnets. So only that piece is missing. I will push it in a subsequent PR. Its not required right now.

Comment (mavenugo) [inline drivers/overlay/ov_network.go]: ok got it.

File: drivers/overlay/joinleave.go (hunk 2 above)
Comment (sanimej) [inline drivers/overlay/joinleave.go]: @mrjana peerKey structure has only the peerIP. I added the mask to peerEntry. That is because on a neighbor miss notification we only get the IP. So in the peerDB we can't change it to IPNet. It can be changed in the serf exchange. But it doesn't buy us much. WDYT ?

I pushed the changes with other comments addressed. PTAL

**Comment by mrjana** [comment]
@sanimej Thanks for taking care of the comments. We can make `peerKey` an `IPNet` and derive the mask for miss-notification from getMatchingSubnet() api but I think we can do that cleanup later. This diff as it is LGTM

**Comment by mavenugo** [comment]
thanks @sanimej . LGTM as well.
can you please open an issue and track the changes that we decided to take up in a subsequent PR ?

**Comment by sanimej** [comment]
@mavenugo Opened #592 to track these changes.

**Comment by mavenugo** [comment]
thanks @sanimej. How about 

>> We can make peerKey an IPNet and derive the mask for miss-notification from getMatchingSubnet() api but I think we can do that cleanup later.

Should we track this ?

**Comment by sanimej** [comment]
This is not a functionality change. I will do it as part of #592
